### PR TITLE
Refactor install/build guide. 

### LINKED ISSUE #1895: [DOC] Refactor and update build/install guide
@nwstephens has pointed out that our build guide has gotten both out of date and hard to follow. In addition, the examples and suggestions in the readme and build/install guide have diverged a bit.  These definitely need some TLC. 

We should probably remove the C++ build without cmake bits as more recent updates have made that near impossible to use in practice. We should also include everything in the build/install guide and then summarize a few bits directly in the readme instead of the other way around.

### PR DESCRIPTION
**Pull request by cjnolet**
This PR makes some mcuh needed changes to the installation and builde guide. This PR also fixes a few APIs that were missing from the API docs. 

Closes #1895

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -255,106 +255,54 @@ pairwise_distance(in1, in2, out=output, metric="euclidean")
...
+Note that the above commands will also install `libraft-headers` and `libraft`.
+
+You can also install the conda packages individually using the `mamba` command above. For example, if you'd like to install RAFT's headers and pre-compiled shared library to use in your project:
+```bash
+# for CUDA 12.0
+mamba install -c rapidsai -c conda-forge -c nvidia libraft libraft-headers cuda-version=12.0
+```
 
-After installing RAFT, `find_package(raft COMPONENTS compiled distributed)` can be used in your CUDA/C++ cmake build to compile and/or link against needed dependencies in your raft target. `COMPONENTS` are optional and will depend on the packages installed.
+If installing the C++ APIs Please see [using libraft](https://docs.rapids.ai/api/raft/nightly/using_libraft/) for more information on using the pre-compiled shared library. You can also refer to the [example C++ template project](https://github.com/rapidsai/raft/tree/branch-23.12/cpp/template) for a ready-to-go Cmake configuration that you can drop into your project and build against installed RAFT development artifacts above.
```
Comment (bdice) [inline README.md:290]:
```suggestion
If installing the C++ APIs please see [using libraft](https://docs.rapids.ai/api/raft/nightly/using_libraft/) for more information on using the pre-compiled shared library. You can also refer to the [example C++ template project](https://github.com/rapidsai/raft/tree/branch-23.12/cpp/template) for a ready-to-go CMake configuration that you can drop into your project and build against installed RAFT development artifacts above.
```

File: README.md [hunk 2]
```diff
@@ -255,106 +255,54 @@ pairwise_distance(in1, in2, out=output, metric="euclidean")
...
 
-### Pip
+### Installing Python through Pip
 
-pylibraft and raft-dask both have experimental packages that can be [installed through pip](https://rapids.ai/pip.html#install):
+`pylibraft` and `raft-dask` both have experimental packages that can be [installed through pip](https://rapids.ai/pip.html#install):
 ```bash
 pip install pylibraft-cu11 --extra-index-url=https://pypi.nvidia.com
```
Comment (bdice) [inline README.md:296]: Do we want to give CUDA 12 instructions (-cu12 suffixes) as well?

File: README.md [hunk 3]
```diff
@@ -255,106 +255,54 @@ pairwise_distance(in1, in2, out=output, metric="euclidean")
...
 pip install raft-dask-cu11 --extra-index-url=https://pypi.nvidia.com
 ```
 
-### CMake & CPM
+These packages statically build RAFT's pre-compiled instantiations and so the C++ headers and pre-compiled shared library won't be readily available to use in your code.
 
-RAFT uses the [RAPIDS-CMake](https://github.com/rapidsai/rapids-cmake) library, which makes it easy to include in downstream cmake projects. RAPIDS-CMake provides a convenience layer around CPM. Please refer to [these instructions](https://github.com/rapidsai/rapids-cmake#installation) to install and use rapids-cmake in your project.
-
-#### Example Template Project
+The [build](https://docs.rapids.ai/api/raft/nightly/build/) instructions contain more details on building RAFT from source and including it in downstream projects. You can also find a more comprehensive version of the above CPM code snippet the [Building RAFT C++ and Python from source](https://docs.rapids.ai/api/raft/nightly/build/#building-c-and-python-from-source) section of the build instructions.
```
Comment (bdice) [inline README.md:302]: I prefer linking both words rather than only one word.
```suggestion
The [build instructions](https://docs.rapids.ai/api/raft/nightly/build/) contain more details on building RAFT from source and including it in downstream projects. You can also find a more comprehensive version of the above CPM code snippet the [Building RAFT C++ and Python from source](https://docs.rapids.ai/api/raft/nightly/build/#building-c-and-python-from-source) section of the build instructions.
```

File: README.md [hunk 4]
```diff
@@ -255,106 +255,54 @@ pairwise_distance(in1, in2, out=output, metric="euclidean")
...
 
 You can find an [example RAFT](cpp/template/README.md) project template in the `cpp/template` directory, which demonstrates how to build a new application with RAFT or incorporate RAFT into an existing cmake project.
```
Comment (bdice) [inline README.md:304]: I'd link "RAFT project template".
```suggestion
You can find an example [RAFT project template](cpp/template/README.md) in the `cpp/template` directory, which demonstrates how to build a new application with RAFT or incorporate RAFT into an existing CMake project.
```

File: docs/source/build.md [hunk 5]
```diff
@@ -1,12 +1,41 @@
 # Installation
 
-### Conda
+RAFT currently provides libraries for C++ and Python. The C++ libraries, including the header-only and optional shared library, can be installed with Conda. 
+
+Both the C++ and Python APIs require cmake to build from source.
```
Comment (bdice) [inline docs/source/build.md:5]:
```suggestion
Both the C++ and Python APIs require CMake to build from source.
```

File: docs/source/build.md [hunk 6]
```diff
@@ -19,19 +48,27 @@ mamba install -c rapidsai -c conda-forge -c nvidia raft-dask pylibraft cuda-vers
...
 
-After installing RAFT, `find_package(raft COMPONENTS nn distance)` can be used in your CUDA/C++ cmake build to compile and/or link against needed dependencies in your raft target. `COMPONENTS` are optional and will depend on the packages installed.
+You can also install the conda packages individually using the `mamba` command above. For example, if you'd like to install RAFT's headers and pre-compiled shared library to use in your project:
+```bash
+# for CUDA 12.0
+mamba install -c rapidsai -c conda-forge -c nvidia libraft libraft-headers cuda-version=12.0
+```
 
-### Pip
+If installing the C++ APIs Please see [using libraft](https://docs.rapids.ai/api/raft/nightly/using_libraft/) for more information on using the pre-compiled shared library. You can also refer to the [example C++ template project](https://github.com/rapidsai/raft/tree/branch-23.12/cpp/template) for a ready-to-go Cmake configuration that you can drop into your project and build against installed RAFT development artifacts above.
```
Comment (bdice) [inline docs/source/build.md:59]:
```suggestion
If installing the C++ APIs Please see [using libraft](https://docs.rapids.ai/api/raft/nightly/using_libraft/) for more information on using the pre-compiled shared library. You can also refer to the [example C++ template project](https://github.com/rapidsai/raft/tree/branch-23.12/cpp/template) for a ready-to-go CMake configuration that you can drop into your project and build against installed RAFT development artifacts above.
```

File: docs/source/build.md [hunk 7]
```diff
@@ -19,19 +48,27 @@ mamba install -c rapidsai -c conda-forge -c nvidia raft-dask pylibraft cuda-vers
...
 
-pylibraft and raft-dask both have experimental packages that can be [installed through pip](https://rapids.ai/pip.html#install):
+## Installing Python through Pip
+
+`pylibraft` and `raft-dask` both have experimental packages that can be [installed through pip](https://rapids.ai/pip.html#install):
```
Comment (bdice) [inline docs/source/build.md:63]: Are these still experimental?
```suggestion
`pylibraft` and `raft-dask` both have packages that can be [installed through pip](https://rapids.ai/pip.html#install):
```

File: docs/source/build.md [hunk 8]
```diff
@@ -19,19 +48,27 @@ mamba install -c rapidsai -c conda-forge -c nvidia raft-dask pylibraft cuda-vers
...
 ```bash
 pip install pylibraft-cu11 --extra-index-url=https://pypi.nvidia.com
 pip install raft-dask-cu11 --extra-index-url=https://pypi.nvidia.com
 ```
 
-## Building and installing RAFT
+These packages statically build RAFT's pre-compiled instantiations and so the C++ headers and pre-compiled shared library won't be readily available to use in your code.
```
Comment (bdice) [inline docs/source/build.md:69]:
```suggestion
These packages statically build RAFT's pre-compiled instantiations, so the C++ headers and pre-compiled shared library won't be readily available to use in your code. 
```

File: docs/source/build.md [hunk 9]
```diff
@@ -57,9 +94,17 @@ In addition to the libraries included with cudatoolkit 11.0+, there are some oth
 - [Googlebench](https://github.com/google/benchmark) - Needed to build benchmarks
 - [Doxygen](https://github.com/doxygen/doxygen) - Needed to build docs
 
-All of RAFT's C++ APIs can be used header-only but pre-compiled shared libraries also contain some host-accessible APIs and template instantiations to accelerate compile times.
+#### Conda environment scripts
+
+Conda environment scripts are provided for installing the necessary dependencies to build both the C++ and Python libraries from source. It is preferred to use `mamba`, as it provides significant speedup over `conda`. In addition you will have to manually install a CUDA toolkit which is greater than, or equal to, the version you install into you conda environment. Installing CUDA toolkit into your host system is necessary because `nvcc` is not provided with Conda's cudatoolkit dependencies. The following example will install create and install dependencies for a CUDA 11.8 conda environment:
```
Comment (bdice) [inline docs/source/build.md:99]: This situation changes pretty dramatically (for the better) with CUDA 12. Should we figure out a way to tell the "simple" CUDA 12 story first, which has no system CTK requirement, and then describe the CUDA 11 case after that? (Eventually the CUDA 11 info will be safe to remove, so I'd try to separate it now.)

File: docs/source/build.md [hunk 10]
```diff
@@ -57,9 +94,17 @@ In addition to the libraries included with cudatoolkit 11.0+, there are some oth
...
+```bash
+mamba env create --name rapids_raft -f conda/environments/all_cuda-118_arch-x86_64.yaml
+mamba activate rapids_raft
+```
+
+All of RAFT's C++ APIs can be used header-only and optional pre-compiled shared libraries provide some host-accessible runtime APIs and template instantiations to accelerate compile times.
 
-The recommended way to build and install RAFT is to use the `build.sh` script in the root of the repository. This script can build both the C++ and Python artifacts and provides options for building and installing the headers, tests, benchmarks, and individual shared libraries.
+The recommended way to build and install RAFT from source is to use the `build.sh` script in the root of the repository. This script can build both the C++ and Python artifacts and provides Cmake options for building and installing the headers, tests, benchmarks, and the pre-compiled shared library.
```
Comment (bdice) [inline docs/source/build.md:107]:
```suggestion
The recommended way to build and install RAFT from source is to use the `build.sh` script in the root of the repository. This script can build both the C++ and Python artifacts and provides CMake options for building and installing the headers, tests, benchmarks, and the pre-compiled shared library.
```

File: docs/source/build.md [hunk 11]
```diff
@@ -206,7 +212,7 @@ python setup.py build_ext --inplace
 python setup.py install
 ```
 
-To run the Python tests:
+The python tests are automatically installed with the Python libraries. To run the Python tests:
```
Comment (bdice) [inline docs/source/build.md:215]:
```suggestion
Python tests are automatically installed with the corresponding libraries. To run Python tests:
```

File: docs/source/build.md [hunk 12]
```diff
@@ -220,27 +226,56 @@ The Python packages can also be uninstalled using the `build.sh` script:
 ./build.sh pylibraft raft-dask --uninstall
 ```
 
-### Documentation
+### Using Cmake directly
```
Comment (bdice) [inline docs/source/build.md:229]: Let's replace all "Cmake" or "cmake" with "CMake" unless referring to the executable `cmake`. Just want to avoid leaving a comment for every instance to aid with readability of the feedback. :)
```suggestion
### Using CMake directly
```

File: docs/source/build.md [hunk 13]
```diff
@@ -220,27 +226,56 @@ The Python packages can also be uninstalled using the `build.sh` script:
...
+| RAFT_COMPILE_LIBRARY            | ON, OFF              | ON if either BUILD_TESTS or BUILD_PRIMS_BENCH is ON; otherwise OFF | Compiles all `libraft` shared libraries (these are required for Googletests) |
+| RAFT_ENABLE_CUBLAS_DEPENDENCY   | ON, OFF | ON | Link against cublas library in `raft::raft`                                  | 
+| RAFT_ENABLE_CUSOLVER_DEPENDENCY | ON, OFF | ON | Link against cusolver library in `raft::raft`                                | 
+| RAFT_ENABLE_CUSPARSE_DEPENDENCY | ON, OFF | ON | Link against cusparse library in `raft::raft`                                | 
+| RAFT_ENABLE_CUSOLVER_DEPENDENCY | ON, OFF | ON | Link against curand library in `raft::raft`                                  | 
+| RAFT_NVTX                       | ON, OFF              | OFF | Enable NVTX Markers                                                          |
+
+### Build documentation
+
+The documentation requires that the C++ and python libraries have been built and installed. The following will build the docs along with the C++ and Python packages:
```
Comment (bdice) [inline docs/source/build.md:263]: Similarly for "Python" instead of "python" throughout.
```suggestion
The documentation requires that the C++ and Python libraries have been built and installed. The following will build the docs along with the C++ and Python packages:
```

**Review by bdice — COMMENTED** [review]
These are helpful updates! Thanks. I left a few high-level comments but please treat my feedback as non-blocking since I may not be able to re-review in a timely fashion.

File: docs/source/build.md [hunk 14]
```diff
@@ -19,19 +48,35 @@ mamba install -c rapidsai -c conda-forge -c nvidia raft-dask pylibraft cuda-vers
...
 
-pylibraft and raft-dask both have experimental packages that can be [installed through pip](https://rapids.ai/pip.html#install):
+## Installing Python through Pip
+
+`pylibraft` and `raft-dask` both have packages that can be [installed through pip](https://rapids.ai/pip.html#install). 
+
+For CUDA 11 packages:
 ```bash
 pip install pylibraft-cu11 --extra-index-url=https://pypi.nvidia.com
 pip install raft-dask-cu11 --extra-index-url=https://pypi.nvidia.com
```
Comment (bdice) [inline docs/source/build.md:68]: Maybe combine?
```suggestion
pip install pylibraft-cu11 raft-dask-cu11 --extra-index-url=https://pypi.nvidia.com
```

File: docs/source/build.md [hunk 15]
```diff
@@ -220,27 +240,56 @@ The Python packages can also be uninstalled using the `build.sh` script:
...
 
+## Using RAFT C++ in downstream projects
 
-## Using RAFT in downstream projects
+There are a few different strategies for including RAFT in downstream projects, depending on whether the [required build dependencies](#build-dependencies) have already been installed and are available on the `lib` and `include` search paths.
 
-There are a few different strategies for including RAFT in downstream projects, depending on whether the [required build dependencies](#build-dependencies) have already been installed and are available on the `lib` and `include` paths.
-
-Using cmake, you can enable CUDA support right in your project's declaration:
+When using the GPU parts of RAFT, you will need to enable CUDA support in your Cmake project declaration:
```
Comment (bdice) [inline docs/source/build.md:287]: I didn't mark all these as noted in my previous review -- maybe use find/replace "Cmake" to "CMake."
```suggestion
When using the GPU parts of RAFT, you will need to enable CUDA support in your CMake project declaration:
```

**Review by bdice — APPROVED** [review]
Approving with a couple minor comments. The changes are a big improvement!

**Review by dantegd — APPROVED** [review]
changes look great!
